VACMS-13769 Fix non page facilities urls pushed to lighthouse

swirtSJW opened the pull request:
## Description

Closes #13769

Summary:  

- Adding url pushes needs to make sure we push when the url would change.
- Adding pushes for Vet Center mobile, which we never pushed before because they have no status.
- Reducing the number of testing permutations while adding more data fields that need to be monitored for change.
- Significant refactoring to allow for different status vs non-status push classes, some shared functions that can be used across them, both in a base class and in a static FacilityOps class.

## Testing done


## Screenshots


## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  

As user admin.
Test an edit of an existing node for one of each facility content types. Changing data is not needed.
1.  PUBLISH a save to  '[health_care_local_facility](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/1816/edit)',
- [x] Validate that a message appears saying ut was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link to the va.gov facility page
2. Save a DRAFT of an '[nca_facility](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/56017/edit)',
- [x] Validate that a message appears saying it was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link URL with a value of NULL
3. Save a DRAFT if a '[vba_facility](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/48912/edit)',
- [x] Validate that a message appears saying it was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link a link URL with a value of NULL
4. Save a DRAFT and a PUBLISH '[vet_center_cap](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/35952/edit)',
- [x]  validate that NO message appears saying it was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that there is NO item queued for the CAP, because Caps are not being pushed.
5.  Save and PUBLISH a published '[vet_center_mobile_vet_center](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/17518/edit)',
- [x] Validate that a message appears saying it was queued for push to facility locator
- [ ] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link to the facility locations page with a hash(fragment) link down to the heading for this facility.
6. save a and PUBLISH a published '[vet_center_outstation](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/17529/edit)',
- [x] Validate that a message appears saying it was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link to the facility locations page with a hash link down to the heading for this facility.
7. save a and PUBLISH a published '[vet_center](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/3718/edit)',
- [x] Validate that a message appears saying ut was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link to the va.gov facility page
1.  save a draft ato a NEW   '[health_care_local_facility](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/node/50479/edit)' that has not been published
- [x] Validate that a message appears saying it was queued for push to facility locator
- [x] Go to [the queue](https://pr13814-03njwdfrcsvv97vua1ehc5mq1nutbus1.ci.cms.va.gov/admin/config/post-api/queue) and validate that the payload for this node entry appears with a link to the Facility Locator page page

swirtSJW commented:
I am sure some of my refactoring is going to anger the  tests.  I will address those in the morning.

In the tests, they need to allow for pushing when archieved or published (event without changes)

Also need to allow pushing with parent or name change.

swirtSJW commented:
This has been a pretty heavy refactoring of how we analyze changes in the values we are pushing and should allow me to simplify the tests and not have so many perutations tests without losing coverage.

omahane reviewed (commented):
I just saw a few things that gave me pause. This is great refactoring and new functionality.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/FacilityOps.php:
```diff
@@ -0,0 +1,132 @@
+   *   TRUE if it is a facility with status info. FALSE otherwise.
+   */
+  public static function isFacility(NodeInterface $node) : bool {
+    $facilities_with_status = [
```
Right.  VetCenter Cap is not here yet because it does not sync with LH yet.  That is  a different PR that will come later.

swirtSJW replied:
I took care of making it true for CAPS.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/FacilityOps.php:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  /**
+   * Checks if the entity is a facility node with status info.
```
Good catch.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/FacilityOps.php:
```diff
@@ -0,0 +1,132 @@
+   * @return bool
+   *   TRUE if it is a facility bundle with status info. FALSE otherwise.
+   */
+  public static function isBundleFacilityWithStatus(string $type) : bool {
```
It is but not yet because lighthouse is not ready to recieve them AND I am not sure if we are going to push them with this since CAPS will have a completely different endpoint.   It is weird because they are a product that has launched but is not ready for LH yet.  This is a good callout though because from a pure FacilityOps point of view that is not concerned about LH.   I will adjust this... have to fix in two places.

swirtSJW replied:
I took care of this.

swirtSJW commented on docroot/modules/custom/va_gov_facilities/src/FacilityOps.php:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  /**
+   * Check if the entity is a facility is a product that has launched.
```
Nice

swirtSJW commented:
These failing tests are a nightmare.  Looks like I am taking this one with me on vacation.  I should have a PR waiting by Monday morning.

swirtSJW commented:
@omahaneI did a lot of refactoring from the Badlands:
- Refactored test to only test that pushes would be performed.  Removing aspects looking at what changed
- Added an interface so that we could define a couple of required public static methods that can address whether something would be pushed.  Making them static meant I could remove a ton of the mocking logic in the test.
-  Consolidated the concrns of many of the functions by moving them to their more appropriate classes.
There is going to be a bunch of testing steps  which I will try to have written up shortly.

It is important this goes out soon.  
-

swirtSJW commented:
Oh bother... those failing phpunit tests seem to be unrelated but will re-run them with fingers crossed.

swirtSJW commented on docroot/modules/custom/va_gov_post_api/src/Service/PostFacilityBase.php:
```diff
@@ -352,4 +325,120 @@ protected function stringNullify($string) {
+   * @return bool
+   *   TRUE if the value changed.  FALSE otherwise.
+   */
+  protected function changedValue(NodeInterface $node, NodeInterface $revision, $field_name): bool {
```
Not used at the moment.  but will be.

swirtSJW commented on docroot/modules/custom/va_gov_post_api/src/Service/PostFacilityBase.php:
```diff
@@ -352,4 +325,120 @@ protected function stringNullify($string) {
+   * @return bool
+   *   TRUE if the value changed.  FALSE otherwise.
+   */
+  protected function changedTarget(NodeInterface $node, NodeInterface $revision, $field_name): bool {
```
Not used at the moment.  but will be.

swirtSJW commented on docroot/modules/custom/va_gov_post_api/src/Service/PostFacilityBase.php:
```diff
@@ -352,4 +325,120 @@ protected function stringNullify($string) {
+   * @return bool
+   *   TRUE if there have been changes, FALSE otherwise.
+   */
+  protected function fieldsHaveChanges(NodeInterface $current_revision, NodeInterface $default_revision, array $field_names): bool {
```
Not used at the moment.  but will be.

swirtSJW commented on docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:
```diff
@@ -6,50 +6,52 @@
-  _post_api_add_facility_status_to_queue($entity);
-  _post_api_add_facility_service_to_queue($entity);
-
+function va_gov_post_api_entity_insert(EntityInterface $entity) {
```
THis is why we have to get this out.  Apparently when I merged Jays work in last sprint his work pre-dated rename-spacing the module and I did not catch the name change getting reverted here.

swirtSJW commented:
Something is up with these phpunit failures.  As near as I can tell these are unrelated to the changes I made to the tests.

```
Run bin/phpunit \ --group unit \ --exclude-group disabled \ --coverage-text \ tests/phpunit
PHPUnit 9.6.[8](https://github.com/department-of-veterans-affairs/va.gov-cms/actions/runs/5108632673/jobs/9182628123#step:4:9) by Sebastian Bergmann and contributors.

Testing /home/runner/work/va.gov-cms/va.gov-cms/tests/phpunit
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
PHP Deprecated:  strtr(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/va.gov-cms/va.gov-cms/docroot/core/lib/Drupal/Component/Render/FormattableMarkup.php on line 257
.......WWWWWWWWWWW.............................................  63 / 288 ( 21%)
............................................................... 126 / 288 ( 43%)
............................................................... 18[9](https://github.com/department-of-veterans-affairs/va.gov-cms/actions/runs/5108632673/jobs/9182628123#step:4:10) / 288 ( 65%)
............................................................... 252 / 288 ( 87%)
................................WWWW                            288 / 288 ([10](https://github.com/department-of-veterans-affairs/va.gov-cms/actions/runs/5108632673/jobs/9182628123#step:4:11)0%)

Time: 00:06.[12](https://github.com/department-of-veterans-affairs/va.gov-cms/actions/runs/5108632673/jobs/9182628123#step:4:13)6, Memory: 50.00 MB

There were 15 warnings:

1) tests\phpunit\Content\NodeLinkEnforcementFilterTest::testProcess with data set #0 ('Some innocuous text with some... text.')
"@covers NodeLinkEnforcementFilter::process" is invalid

/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/Framework/TestResult.php:853
/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:651
/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:[14](https://github.com/department-of-veterans-affairs/va.gov-cms/actions/runs/5108632673/jobs/9182628123#step:4:15)4
/home/runner/work/va.gov-cms/va.gov-cms/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:97
```

I can not reproduce these locally (though admitttedly I am working on my chromebook and not my dailytdriver ) .  I feel like I am missing something stupid with these, @ndouglas any chance you can take a quick look and tell me if I a missing a legit failure on my tests?

ndouglas commented:
@swirtSJW You're looking at a warning in the unit tests, but what actually failed is a functional test:

```
PHPUnit 9.6.8 by Sebastian Bergmann and contributors.

Testing /var/lib/tugboat/tests/phpunit
R
Deprecated: Return type of GuzzleHttp\Cookie\CookieJar::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/lib/tugboat/docroot/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php on line 220

Deprecated: Return type of GuzzleHttp\Cookie\CookieJar::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/lib/tugboat/docroot/vendor/guzzlehttp/guzzle/src/Cookie/CookieJar.php on line 225
.......ESSS..............................S....................  63 / 173 ( 36%)
...........................................SSS................. 126 / 173 ( 72%)
...............................................                 173 / 173 (100%)

Time: 04:33.357, Memory: 3.44 GB

There was 1 error:

1) tests\phpunit\Content\AliasesTest::testVamcFacilityHealthServiceAlias
Error: Call to undefined method Drupal\va_gov_post_api\Service\PostFacilityService::isLovellTricareSection()

/var/lib/tugboat/docroot/modules/custom/va_gov_post_api/src/Service/PostFacilityService.php:349
/var/lib/tugboat/docroot/modules/custom/va_gov_post_api/src/Service/PostFacilityService.php:172
/var/lib/tugboat/docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:77
/var/lib/tugboat/docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:20
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Extension/ModuleHandler.php:426
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Extension/ModuleHandler.php:405
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Extension/ModuleHandler.php:433
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/EntityStorageBase.php:251
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/ContentEntityStorageBase.php:903
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/EntityStorageBase.php:598
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/ContentEntityStorageBase.php:784
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/EntityStorageBase.php:523
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php:804
/var/lib/tugboat/docroot/core/lib/Drupal/Core/Entity/EntityBase.php:339
/var/lib/tugboat/docroot/core/modules/node/tests/src/Traits/NodeCreationTrait.php:105
/var/lib/tugboat/docroot/vendor/weitzman/drupal-test-traits/src/Entity/NodeCreationTrait.php:24
/var/lib/tugboat/tests/phpunit/Content/AliasesTest.php:72
/var/lib/tugboat/docroot/vendor/phpunit/phpunit/src/Framework/TestResult.php:728
/var/lib/tugboat/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/var/lib/tugboat/docroot/vendor/phpunit/phpunit/src/Framework/TestSuite.php:684
/var/lib/tugboat/docroot/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:651
/var/lib/tugboat/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:144
/var/lib/tugboat/docroot/vendor/phpunit/phpunit/src/TextUI/Command.php:97
```

I think that the warning you mention is indeed unrelated to your changes. I'll look at it separately, since it's presumably happening on `main`.

omahane commented:
![Screen Shot 2023-05-30 at 10 50 09 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/766573/8ae58246-6806-4c6f-8df5-3d2f6a081fdb)
